openqaoa/optimizers/result.py

```diff
@@ -143,3 +143,54 @@ def plot_cost(self, figsize=(10, 8), label='Cost', linestyle='--', color='b', ax
         ax.set_title('Cost history')
 
         return
+
+    def plot_probabilities(self, figsize = (10,8),label='Probability distribution',color='tab:blue', ax=None):
+
+        """
+        Helper function to plot the probabilities corresponding to each basis states (with prob != 0) obtained from the optimized result
```

Comment:
We might want to consider having a warning in the docstring when using this method. It took quite some time for a graph from a 8-Qubit computation to be plotted.

Reply:
Very good point! Two thoughts: 

1. If you want to plot something with more than, say, 8 qubits, then the plot contains the most probable states. The actual number should be a fraction that gets smaller as the number of qubits increases. Question: given that you have run a simulation over `25` qubits: how many states should we really want to plot? `2^25` is out of the question. But also, the most probable `1%` could potentially be very large.

2. This method should work also when plotting the results of a QPU with `O(100)` qubits, and the current implementation will just freeze your computer if you try to plot it. Now, let's assume we have 50 qubits. That is `2^50` possible states. You are doing shots, so in the worst case where each shot yields a different state, you only need to plot `#shots`. So, we could implement some general rules for displaying these states without freezing the machine

So, in both cases I think we should have clip the number of states we want to plot and let the user know that we did that and (perhaps) give them the opportunity to overwrite the warning and crash their laptops :D

Reply:
So, should I just choose an arbitrary number of states to show for all the cases or rather a dynamic one? If so, how to choose this number?

Reply:
I think the main issue is not the number of data-points but the labels. So, I propose:
- We cap the number of labels in the plot to something like ~100
- We could include a panda's data frame result where you get get the statistics of larger distributions and where you can subsequently plot them on your own time and desire.

Thoughts, @vishal-ph @shahidee44  @EmilianoG-byte

Reply:
I like what the `qiskit.tools.visualization.plot_histogram` does here. As a user, you have an option to plot the most probable `x` states individually (with their labels displayed on the x-axis). The remaining state probabilities are added in a single bar at the end (with a label, for instance, "remaining states/ other states"). 
Therefore, in total the plotter will only show `x+1` bars on the picture.

Reply:
I think even if we just limit the number of labels that can be the, just trying to plot 2^N bars will not scale very well in the long run. So we would still have to reduce the number of data-points here.

Since we are thinking slightly long term here, I'd say we should have some kind of plot cap. (Maybe "best" 100 states ranked by probability(?)). 

At the same time, I don't think we should overcomplicate some of these methods since we intend to add more methods that could visualize the outputs in various different ways. The second point you propose would probably make sense as a seperate method within the Result object.

Alternatively, we can just produce some kind of error if the user tries to plot to many data-points. Like raising a warning if they try to plot more than 100 data-points.

Reply:
Ok, then I think we can go with what Vishal proposed. 

Sounds fair, @EmilianoG-byte ?

Reply:
Ok. So, if I understood correctly, we will give the option to the user to select the top `x` states, and set a default value for `x` (and equal upper bound) of 100 states. Is that correct?

Reply:
Yes. Then we just sum up the probabilities of the remaining state and add that as a final bar in the graph. Therefore, we would have `x+1` bars in the bar graph. (Maybe you can try to play around with the default value and see whats a good default to set such that the graph is still easily readable too)

Reply:
Is it okay if I also increase the figsize to include more states? So far I was using the same as for the `plot_cost` method which is (10,8). If I do change mine, should I change the one for the cost as well?

Reply:
have done the changes, let me know what you think :)

Reply:
I think it is quite nice! I have just added a few comments.

Also, note that you may need to pull the branch, since I have made a few changes to make sure the tests were all green on our end :)

Reply:
Is anybody else having the error `Pauli terms should be of type PauliOp and not <class 'openqaoa.qaoa_parameters.operators.PauliOp'>` when compiling the problem? This was not happening before

Reply:
Not really, I have tried, and all seems fine. Can you let us know how to reproduce the error?

Reply:
> Ok! Besides the potential issue regarding the Pauli op error, I think this PR is ready to be merged! (cc @vishal-ph )

I see, there are still some changes requested by Shahidee. Are those already implemented? If yes, lets resolve the comments and we can proceed with merge

Reply:
Initially, I got it when compiling a QUBO for ShortPath Problem, but the problem is there for something as simple as:


`problem_qubo = QUBO(3, [[0],[1],[2]], [1,1,1])`
`q = QAOA()`
`q.compile(problem_qubo)`
`q.optimize()`

Reply:
Ok, I just reloaded VS code and everything works fine. I don't what was going on. Sorry for the confusion :/